json_yaml_files/prepare_files.py

```diff
@@ -0,0 +1,178 @@
+            json.dump(json_contents, f)
+
+    def fetch_scan_type(self, Y, Z):
+        scan_type = ''
+        if Z.lower() in self.scan_type[Y]:
+            scan_type = self.scan_type[Y][Z.lower()]
+        elif '_' in Z:
+            first = Z.split('_')[0]
+            if first.lower() in self.scan_type[Y]:
+                scan_type = self.scan_type[Y][first.lower()]
+        if Y in self.scan_type[Y]:
+            scan_type = self.scan_type[Y][Y]
+        return scan_type
+
+    def prepare_yaml_contents(self, file_name):
+        _, _, Y, Z = self.extract_file_name_components(file_name)
+
+        yaml_contents = {
+            "scan_type": self.fetch_scan_type(Y, Z),
+            "scan_object": 'LIVE',
+            "image_file_format": 'NIFTI',
+            "image_modality": self.image_modality[Y],
+            "transformation_performed": 'Yes'
+        }
+        finalPath = '/'.join(self.path.split('/')[:-2]) + '/prepared_yamls'
+        Path(finalPath).mkdir(parents=True, exist_ok=True)
+        with open(f'{finalPath}/'+file_name, 'w') as f:
+            yaml.dump(yaml_contents, f)
+
+    def fetch_Z_value(self, files):
+        Z = set()
+        ignore_list = ['desc', 'subject', 'session',
+                       'extension', 'suffix', 'datatype']
+        for file in files:
+            entities = self.layout.parse_file_entities(file)
+            # print(entities)
+            filtered_entities = {k: v for k,
+                                 v in entities.items() if k not in ignore_list}
+            # print(filtered_entities)
+            for key, value in filtered_entities.items():
+                if key != 'space':
+                    z = key + '-' + value
+                else:
+                    z = value
+                if ({'suffix', 'datatype'} <= entities.keys() and entities['suffix'] != entities['datatype']) or \
+                        ('suffix' in entities and 'datatype' not in entities):
+                    if z:
+                        z = z + '_'
+                    z = z + entities['suffix']
+                if z:
+                    Z.add(z)
+            if not Z and entities['suffix'] != entities['datatype']:
+                Z.add(entities['suffix'])
+        return list(Z)
+
+    def prepare_file_names(self, X, Y, Z):
+        file_names = defaultdict(set)
+        filename = f"{self.A}_{X}"
+        file_base = f"{filename}.{Y}."
+        for z in Z:
+            file_names['json'].add(file_base + z + '.json')
+            file_names['yaml'].add(file_base + z + '.yaml')
+        file_names['json'] = list(file_names['json'])
+        file_names['yaml'] = list(file_names['yaml'])
+        return file_names
+
+    def run(self):
+        X = []
+        Xi = self.layout.get(scope='raw')  # inputs
+        # Xd = self.layout.get(scope='derivatives')  # derivatives
+        Xs = self.layout.get(scope='sourcedata')  # sourcedata
+        if Xi:
+            X.append('inputs')
```

Comment:
Would "inputs" be more accurately renamed as "rawdata"? I don't know PyBIDS well enough to know. This is just a comment question, not a show-stopper.

Reply:
`inputs` is based on the value that OpenNeuro expects in the file name whose type is raw. And to identify that from the BIDS dataset, `scope='raw'` is used.